Enhanced API Client with Advanced Rate Limiting and Memory Management

patricioxavier8 opened the pull request:
This PR complete rewrite of APIClient to handle high-volume file uploads with improved reliability, memory efficiency, and intelligent rate limiting 

Key Improvements

Memory Management & Resource Control

Bounded semaphore with configurable max waiters  to prevent memory exhaustion
Proper task cancellation using dedicated URLSessionTaskActor with weak references
Queue overflow protection - fails fast when request queue is full instead of growing indefinitely

Intelligent Rate Limiting

Adaptive token bucket with burst capacity and configurable refill rate
Server-aware throttling tracks consecutive failures and server-indicated backoff periods
Jittered exponential backoff prevents thundering herd scenarios

Enhanced Retry Logic

-Comprehensive Retry-After parsing supports both numeric seconds and HTTP date formats
-Adaptive delay calculation adjusts based on failure history and server signals

patricioxavier8 committed:
Add rate limiter to control API request frequency

patricioxavier8 committed:
rewrite of APIClient to handle high-volume file uploads with improved reliability, memory efficiency, and intelligent rate limiting

patricioxavier8 committed:
add tests for apiclient

patricioxavier8 committed:
add import

patricioxavier8 committed:
Refactor MockURLProtocol to support async request handlers

sg-gs commented on Sources/InternxtSwiftCore/Services/HttpAPI/APIClient/APIClient.swift:
```diff
@@ -11,116 +11,554 @@ import Combine
         urlRequest.setValue(clientName, forHTTPHeaderField: "internxt-client")
         urlRequest.setValue(clientVersion, forHTTPHeaderField: "internxt-version")
-        
+        urlRequest.setValue("application/json", forHTTPHeaderField: "Content-Type")
```
Be careful, this is not the case when uploading a file. As this is the APIClient, I assume it is used for uploads, or not?

patricioxavier8 replied:
Yes, it is for uploads too, regarding that detail, I have not changed anything, I only changed the position of that line.

sg-gs commented on Tests/InternxtSwiftCoreTests/Services/Network/APIClientTests.swift:
```diff
@@ -0,0 +1,165 @@
+        let endpoint = Endpoint(path: "https://mock.test", method: .GET)
+
+        do {
+            _ = try await client.fetch(type: DummyResponse.self, endpoint)
```
You are not verifying really testing the adaptive delay here, you also need to check that the delays increase exponentially.

sg-gs commented on Tests/InternxtSwiftCoreTests/Services/Network/APIClientTests.swift:
I miss these tests: 
- testTokenBucketRateLimiting
- testConcurrencyLimiting
- testSemaphoreQueueOverflow

patricioxavier8 committed:
add new tests
-testTokenBucketRateLimiting
-testConcurrencyLimiting
-testSemaphoreQueueOverflow
add new helper methods
